lib/tbot/services/application/proxy_service.go

```diff
@@ -0,0 +1,367 @@
+	// but, proxying them forward as normal requests would make it difficult for
+	// us to introduce CONNECT support later without breaking compat.
+	if req.Method == http.MethodConnect {
+		return trace.NotImplemented("proxy does not support CONNECT method")
+	}
+
+	// net/http implements RFC7230 5.3 correctly, and will prefer the host
+	// specified within an absolute-form request-target over the Host header
+	// when setting req.Host - which means we can safely use req.Host here.
+	appName := req.Host
+
+	var appCert *tls.Certificate
+	var err error
+	appCert, err = utils.FnCacheGet(
+		ctx,
+		s.cache,
+		appName,
+		func(ctx context.Context) (*tls.Certificate, error) {
+			s.log.InfoContext(
+				ctx, "Issuing app cert",
+				"app", appName,
+			)
+			cert, _, err := s.issueCert(ctx, appName)
+			return cert, err
+		})
+	if err != nil {
+		return trace.Wrap(err, "fetching certificate")
+	}
+
+	// TODO(noah): We could cache the httpClient itself for each upstream, this
+	// would potentially allow performance improvements by caching connections.
+	transport := &http.Transport{
+		TLSClientConfig: &tls.Config{
+			Certificates:       []tls.Certificate{*appCert},
+			InsecureSkipVerify: s.botClient.Config().InsecureSkipVerify,
+		},
+	}
+	// Inject the ALPN upgrade dialer if required.
+	if s.alpnUpgradeRequired {
+		transport.DialContext = apiclient.NewALPNDialer(apiclient.ALPNDialerConfig{
+			ALPNConnUpgradeRequired: true,
+			TLSConfig: &tls.Config{
+				InsecureSkipVerify: s.botClient.Config().InsecureSkipVerify,
+				NextProtos:         []string{string(common.ProtocolHTTP)},
+			},
+		}).DialContext
+	}
+	httpClient := &http.Client{
+		Transport: transport,
+	}
+
+	// Build the Application Request
+	upstreamReq := req.Clone(req.Context())
+	// For now, we redirect all requests to the proxy web's address. However,
+	// it would be a potential improvement for us to switch to fetching the
+	// public address from the App resource. This would reduce the chance of
+	// breakage if the Proxy is eventually modified to consider SNI. The problem
+	// with doing this is that the public address does not include a port, which
+	// we could guess from the Proxy's public web address.
+	upstreamReq.Host = s.proxyAddr
+	upstreamReq.URL.Host = s.proxyAddr
+	upstreamReq.URL.Scheme = "https"
+	// RequestURI must be empty when making client requests.
+	upstreamReq.RequestURI = ""
+	// TODO: Are there any headers we should override, add, or remove on the
+	// upstream request?
+
+	// Execute the upstream request
+	resp, err := httpClient.Do(upstreamReq)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	// Prepare and send the response back to the client
+	// Copy headers first.
+	for key, values := range resp.Header {
+		for _, value := range values {
+			w.Header().Set(key, value)
+		}
+	}
+	// Send status code and then copy the body.
+	w.WriteHeader(resp.StatusCode)
+	if _, err := io.Copy(w, resp.Body); err != nil {
```

Comment:
Possibly not relevant, but the stdlib's HTTP [reverse proxy implementation](https://go.dev/src/net/http/httputil/reverseproxy.go#L626) seems to have some special logic around copying the response (e.g. flush intervals).

Reply:
As far as I can tell, we'd likely need to add something similar to this if we wanted to support WebSockets or SSE. I've left that out of scope for now since both of the customers I'm thinking of don't leverage these technologies at the moment.